### ISSUE TITLE
Point centre location changes with zoom. 

### ISSUE THREAD

**Issue by sun-jiao** [tprebs/fluttermap_heatmap#7]
[录屏 2023-05-31 14-56-07.webm](https://github.com/tprebs/fluttermap_heatmap/assets/14086980/ae654603-5142-46a7-9e38-b2c97a29ac36)

**Comment by tprebs** [comment]
@sun-jiao, is this just a single point on a map or multiple points? If it's multiple points close together then the data may be being gridded differently at different zoom levels which is why it moves.

**Comment by sun-jiao** [comment]
@tprebs there are two points in `data` but only one is shown in video.

**Comment by epiphanatic** [comment]
@tprebs I've got the same issue. Single points move around when zooming. We are feeding the heatmap marker locations (all spaced very far apart, so they are not aggregating) and when you zoom in an out it moves around, mainly to the bottom right in our case. In any case, the heat map is never centered on the marker.

**Comment by tprebs** [comment]
@epiphanatic I have tried to recreate this but I am unable to so far. Do you have a minimal example project or data set I could use to look into this? What coordinate system are you using?

**Comment by leewrigg** [comment]
@tprebs We're experiencing the same issue - 

This seems to be correlated with the radius size. The heatmap is offset more when the radius is bigger.

The video shows a heatmap and a red 'X' marker on the map. Both of these have the same LatLng point (the Sydney Opera House) so I believe they should be centred on top of each other, however the heatmap is actually offset to the bottom-right (southeast) of the red X. As I zoom in, the heatmap gets closer to the actual point.

https://github.com/tprebs/fluttermap_heatmap/assets/32387991/eecbd748-7cc9-4bda-a4ef-960b020df313

The code for this video example is below. This is modified from the example in the repo:

```
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:flutter_map_heatmap/flutter_map_heatmap.dart';
import 'package:latlong2/latlong.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Heatmap Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'flutter_mapmap heat_map demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  StreamController<void> _rebuildStream = StreamController.broadcast();
  List<WeightedLatLng> data = [];

  initState() {
    _loadData();
    super.initState();
  }

  @override
  dispose() {
    _rebuildStream.close();
    super.dispose();
  }

  final sydneyOperaHouse = LatLng(-33.8568, 151.2153);

  late final redX = MarkerLayer(
    markers: [
      Marker(
          point: sydneyOperaHouse,
          builder: (_) =>
              Center(child: Icon(Icons.close, size: 40, color: Colors.red)))
    ],
  );

  _loadData() async {
    setState(() {
      data = [
        WeightedLatLng(sydneyOperaHouse, 1),
        WeightedLatLng(sydneyOperaHouse, 1),
        WeightedLatLng(sydneyOperaHouse, 1),
        WeightedLatLng(sydneyOperaHouse, 1),
        WeightedLatLng(sydneyOperaHouse, 1),
        WeightedLatLng(sydneyOperaHouse, 1),
      ];
    });
  }

  @override
  Widget build(BuildContext context) {
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      _rebuildStream.add(null);
    });

    final map = new FlutterMap(
      options: new MapOptions(center: sydneyOperaHouse, zoom: 8.0),
      children: [
        TileLayer(
            urlTemplate: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
            subdomains: ['a', 'b', 'c']),
        HeatMapLayer(
          heatMapDataSource: InMemoryHeatMapDataSource(data: data),
          heatMapOptions: HeatMapOptions(
            radius: 100,
          ),
          reset: _rebuildStream.stream,
        ),
        redX, // <--- To show where the heatmap centre should be
      ],
    );
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      backgroundColor: Colors.pink,
      body: Center(
        child: Container(child: map),
      ),
    );
  }
}

```

**Comment by epiphanatic** [comment]
@tprebs ours looks exactly like what @leewrigg has described above. Setup same, effect is the same, increased radius increases to the southeast, etc. 

I haven't looked at your source code in depth yet, but I'm assuming that the heat map is taking the coordinates and then starting the drawing from there and that the drawing function always starts in the 'top left' of whatever you tell it to draw, which would then result in the heat never being centered but appearing that it is getting more centered as you zoom in (since the radius covers less space).

If it is something such as what I just described then you would need to tell the drawing function to offset the starting point in the top left by the radius that is set. That way when it draws the circle the point would be in the middle.

Just a guess but I'm probably not that far off if I am off.

**Comment by tprebs** [comment]
@leewrigg , @epiphanatic  Thanks for the investigation. fixed and released in `0.0.4+1`
Looking back through the code it was offsetting by the width of the base image being rendered for the pixel and not half the width. As I have generally worked with dense datasets and smaller radii I hadn't noticed a drift of a few pixels. Thanks!

**Comment by leewrigg** [comment]
@tprebs Thanks for the quick response, really appreciate the work you've put into this plugin!
